core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -414,6 +433,142 @@ private[spark] object ClosureCleaner extends Logging {
+    val isClosureCandidate = implementedInterfaces.exists(_.getName == "scala.Serializable") &&
+      implementedInterfaces.exists(_.getName.startsWith("scala.Function"))
+
+    if (isClosureCandidate) {
+      try {
+        val lambdaProxy = inspect(maybeClosure)
+        if (isIndylambdaScalaClosure(lambdaProxy)) Option(lambdaProxy)
+        else None
+      } catch {
+        case e: Exception =>
+          // no need to check if debug is enabled here the Spark logging api covers this.
+          logDebug("The given reference is not an indylambda Scala closure.", e)
+          None
+      }
+    } else {
+      None
+    }
+  }
+
+  def isIndylambdaScalaClosure(lambdaProxy: SerializedLambda): Boolean = {
+    lambdaProxy.getImplMethodKind == MethodHandleInfo.REF_invokeStatic &&
+      lambdaProxy.getImplMethodName.contains("$anonfun$")
+      // && implements a scala.runtime.java8 functional interface
+  }
+
+  def inspect(closure: AnyRef): SerializedLambda = {
+    val writeReplace = closure.getClass.getDeclaredMethod("writeReplace")
+    writeReplace.setAccessible(true)
+    writeReplace.invoke(closure).asInstanceOf[SerializedLambda]
+  }
+
+  /**
+   * Scans an indylambda Scala closure, along with its lexically nested closures, and populate
+   * the accessed fields info on which fields on the outer object are accessed.
+   */
+  def findAccessedFields(
+      lambdaProxy: SerializedLambda,
+      lambdaClassLoader: ClassLoader,
+      accessedFields: Map[Class[_], Set[String]]): Unit = {
+    val implClassInternalName = lambdaProxy.getImplClass
+    // scalastyle:off classforname
+    val implClass = Class.forName(
+      implClassInternalName.replace('/', '.'), false, lambdaClassLoader)
+    // scalastyle:on classforname
+    val implClassNode = new ClassNode()
+    val implClassReader = ClosureCleaner.getClassReader(implClass)
+    implClassReader.accept(implClassNode, 0)
+
+    val methodsByName = Map.empty[MethodIdentifier[_], MethodNode]
+    for (m <- implClassNode.methods.asScala) {
+      methodsByName(MethodIdentifier(implClass, m.name, m.desc)) = m
+    }
+
+    val implMethodId = MethodIdentifier(
+      implClass, lambdaProxy.getImplMethodName, lambdaProxy.getImplMethodSignature)
+
+    val visited = Set[MethodIdentifier[_]](implMethodId)
+    val stack = Stack[MethodIdentifier[_]](implMethodId)
+    while (!stack.isEmpty) {
+      val currentId = stack.pop
+      val currentMethodNode = methodsByName(currentId)
+      logTrace(s"  scanning $currentId")
+      currentMethodNode.accept(new MethodVisitor(ASM7) {
+        override def visitFieldInsn(op: Int, owner: String, name: String, desc: String): Unit = {
+          if (op == GETFIELD || op == PUTFIELD) {
+            val ownerExternalName = owner.replace('/', '.')
+            for (cl <- accessedFields.keys if cl.getName == ownerExternalName) {
+              logTrace(s"    found field access $name on $owner")
```

Comment:
Will logging `ownerExternalName` be more readable?

Reply:
Indeed! Thanks, I'll address this in an update.